enhancements/storage/csi-vsphere-operator.md

```diff
@@ -0,0 +1,172 @@
+2. A service account will be created for running the operator.
+3. The operator will get RBAC rules necessary for running the operator and its operand (i.e the actual vSphere CSI driver).
+4. A deployment will be created for starting the operator and will be managed by cluster-storage-operator.
+5. A instance of `ClusterCSIDriver` will be created to faciliate managment of driver operator.  `ClusterCSIDriver` is already defined in - https://github.com/openshift/api/blob/master/operator/v1/types_csi_cluster_driver.go but needs to be expanded to include vSphere CSI driver.
+6. cluster-storage-operator will create required cloud-credentials for talking with vCenter API.
+
+#### Driver deployment via vsphere-csi-driver-operator.
+
+The operator itself will be responsible for running the driver and all the required sidecars (attacher, provisioner etc).
+
+1. The operator will create namespace `openshift-cluster-csi-drivers` to deploy the driver and sidecars.
+2. A service account will be created for running the driver.
+3. The operator will create RBAC rules necessary for running the driver and sidecars.
+4. A deployment will be created and managed by operator to handle control-plane sidecars and controller-plane driver deployment with controller-side of CSI services.
+5. A DaemonSet will be created and managed by operator to run node side of driver operations.
+6. A `CSIDriver` object will be created to expose driver features to control-plane.
+7. The driver operator will use and expose cloud-credentials created by cluster-storage-operator.
+
+Most of the steps outlined above is common to all CSI driver operator and vSphere CSI driver is not unique among those aspects.
+
+#### Additional consideration for vSphere
+
+There are certain aspects of driver which require special handling:
+
+##### StoragePolicy configuration
+
+Currently while deploying Openshift a user can configure datastore used by OCP via install-config.yaml. vSphere CSI driver however can't use datastore directly and must be configured with vSphere storage policy. 
+
+To solve this problem vsphere CSI operator is going to create a storagePolicy by tagging selected datastore in the installer. This will require OCP to have expanded permissions of creating storagePolicies.
+
+If CSI operator can not create storage Policy for some reason:
+
+- The operator will continue installing CSI driver without creating storageClass. The operation of creating storagePolicy and storageClass will be however retried expecting permissions to be fixed or transient errors to go away.
+- It is expected that user will manually create a storageClass to consume vSphere CSI volumes if CSI operator can not create storagePolicy and storageClass.
+- CSI operator will mark itself as degraded but CSO itself will not be degraded and vSphere CSI operator's degraded status must be handled in a different way in CSO.
+- Create a metric if storagePolicy and storageClass creation fails in 4.7.
```

Comment:
Can something else, like the installer, create the storagePolicy? If the operator could assume the storagePolicy exists, it wouldn't have to do any juggling with conditions nor potentially leave the cluster without a default StorageClass.

If creating the storagePolicy somewhere else is **not** an option:

The ultimate goal of the `vsphere-csi-driver-operator` should be _creating a StorageClass in the cluster that users can use to provision their volumes_. If for some reason it lacks any permissions to achieve such a goal, IMO it shouldn't leave the cluster with part of the work done and expect the user to fill in the gaps.

IMO it's reasonable to expect the operator to have such permissions in the cluster.

Reply:
We discussed this in one of the design calls. During an upgrade, there is no installer pod and hence it must be vsphere CSI operator that creates the storagePolicy.

The operator is not going to leave just create storagePolicy, it will also create storageClass. I will update the wording.